Test/SearchTaskTest.cs

```diff
@@ -241,5 +242,53 @@ public static void PrunedDbWithContaminantsTest()
             Assert.That(File.ReadAllLines(Path.Combine(folderPath, @"DbForPrunedDbPruned.xml")).Length > 0);
             Directory.Delete(folderPath, true);
         }
+
+        /// <summary>
+        /// Test ensures pruned databases are written when contaminant DB is searched
+        /// </summary>
+        [Test]
+        public static void PeptideFDRTest()
+        {
+            SearchTask searchTask = new SearchTask()
+            {
+                SearchParameters = new SearchParameters
+                {
+                    WritePrunedDatabase = true
+                },
+            };
+
+            string myFile = Path.Combine(TestContext.CurrentContext.TestDirectory, @"TestData\PrunedDbSpectra.mzml");
+            string myFile2 = Path.Combine(TestContext.CurrentContext.TestDirectory, @"TestData\PrunedDbSpectra2.mzml");
+            if (!File.Exists(myFile2)) { File.Copy(myFile, myFile2); }
+            string myDatabase = Path.Combine(TestContext.CurrentContext.TestDirectory, @"TestData\DbForPrunedDb.fasta");
+            string folderPath = Path.Combine(TestContext.CurrentContext.TestDirectory, @"TestPeptideFDR");
+            string filePath = Path.Combine(TestContext.CurrentContext.TestDirectory, @"TestData\ExperimentalDesign.tsv");
+            DbForTask db = new DbForTask(myDatabase, true);
+            Directory.CreateDirectory(folderPath);
+
+            // search something with multiple hits of the same peptide to see if peptide FDR is calculated at the end
+            searchTask.RunTask(folderPath, new List<DbForTask> { db }, new List<string> { myFile, myFile2 }, "normal");
+
+            List<string> columns = null;
+            int cumDecoys = 0;
+            int cumTargets = 0;
+            foreach (string line in File.ReadAllLines(Path.Combine(folderPath, @"AllPeptides.psmtsv")))
+            {
+                string[] lineline = line.Split('\t');
+                if (line.StartsWith("File Name")) // header
+                {
+                    columns = lineline.ToList();
+                }
+                else if (lineline[columns.IndexOf("Decoy/Contaminant/Target")] == "D")
+                {
+                    Assert.AreEqual(++cumDecoys, int.Parse(lineline[columns.IndexOf("Cumulative Decoy")]));
```

Comment:
I don't really understand this test - you're checking to see if the number of targets/decoys equals the number of targets/decoys if you start counting from the top. isn't that true currently too, even with the buggy FDR? 

I think it would be better to have 3 PSMs, two targets of the same sequence and then the third being a decoy. the q-value of the decoy should be 0.5 if the peptide FDR is calculated incorrectly, and 1.0 if it's calculated correctly, right? q=(targets / decoys)

Reply:
No, it's not true. The target/decoy counts are from PSM FDR analysis. In this test, there are 2 PSMs for every peptide, so with the current code the peptide list goes 1,3,5,7. With these fixes, it goes 1,2,3,4, as expected.

Reply:
I think the Q-value test is reasonable, slightly simpler, and achieves the same result.

On the other hand, this one addresses the counts that are used to calculate the q-value, which were why it became obvious that peptide FDR was broken.

Reply:
ah i see. ok. I do think it's good to have an assertion there that checks the q-value is computed correctly, since that's the measurement we actually care about. can you add that assertion?

Reply:
Sure, I added an assertion for the final q-value. Intermediate ones have assumptions that make them not always (decoy / target)